provider/ec2/instancetypes.go

```diff
@@ -0,0 +1,328 @@
+		return nil, errors.Trace(err)
+	}
+	var zoneNames []string
+	zoneFilter := &ec2.Filter{Name: aws.String("location")}
+	for _, z := range zoneResults.AvailabilityZones {
+		zoneNames = append(zoneNames, *z.ZoneName)
+		zoneFilter.Values = append(zoneFilter.Values, z.ZoneName)
+	}
+
+	// Query the instance type names for the region and credential in use.
+	var instTypeNames []*string
+	instanceTypeRegions := make(map[string]set.Strings)
+	var token *string
+	for {
+		offeringResults, err := ec2Session.DescribeInstanceTypeOfferings(&ec2.DescribeInstanceTypeOfferingsInput{
+			LocationType: aws.String("availability-zone"),
+			MaxResults:   aws.Int64(100),
+			NextToken:    token,
+			Filters:      []*ec2.Filter{zoneFilter},
+		})
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, offering := range offeringResults.InstanceTypeOfferings {
+			if _, ok := instanceTypeRegions[*offering.InstanceType]; !ok {
+				instanceTypeRegions[*offering.InstanceType] = set.NewStrings()
+			}
+			instanceTypeRegions[*offering.InstanceType].Add(*offering.Location)
+			instTypeNames = append(instTypeNames, offering.InstanceType)
+		}
+		token = offeringResults.NextToken
+		if token == nil {
+			break
+		}
+	}
+
+	// Populate the costs fo the instance types in use.
+	costs := make(map[string]uint64)
+	if err := instanceTypeCosts(ec2Session, instTypeNames, zoneNames, costs); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Compose the results.
+	var allInstanceTypes []instances.InstanceType
+	for {
+		querySize := len(instTypeNames)
+		if querySize > 100 {
+			querySize = 100
+		}
+		instTypeParams := &ec2.DescribeInstanceTypesInput{
+			InstanceTypes: instTypeNames[0:querySize],
+		}
+		instTypeResults, err := ec2Session.DescribeInstanceTypes(instTypeParams)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, info := range instTypeResults.InstanceTypes {
+			instType := instances.InstanceType{
+				Name:       *info.InstanceType,
+				VirtType:   virtType(info),
+				Deprecated: info.CurrentGeneration == nil || !*info.CurrentGeneration,
+				Cost:       costs[*info.InstanceType],
+			}
+			if info.VCpuInfo != nil && info.VCpuInfo.DefaultVCpus != nil {
+				instType.CpuCores = uint64(*info.VCpuInfo.DefaultVCpus)
+				if info.ProcessorInfo != nil && info.ProcessorInfo.SustainedClockSpeedInGhz != nil {
+					cpupower := calculateCPUPower(
+						instType.Name,
+						info.ProcessorInfo.SustainedClockSpeedInGhz,
+						uint64(*info.VCpuInfo.DefaultVCpus))
+					instType.CpuPower = &cpupower
+				}
+			}
+			if info.MemoryInfo != nil && info.MemoryInfo.SizeInMiB != nil {
+				instType.Mem = uint64(*info.MemoryInfo.SizeInMiB)
+			}
+			if info.ProcessorInfo != nil {
+				for _, instArch := range info.ProcessorInfo.SupportedArchitectures {
+					instType.Arches = append(instType.Arches, archName(*instArch))
+				}
+			}
+			if instRegions, ok := instanceTypeRegions[instType.Name]; !ok || instRegions.Size() < len(zoneNames) {
```

Comment:
lets change this logic to, if the region has more than 3 zones, then as long as 3 zones support it, allow it, otherwise if the region has less than 3 zones, only support if its supported in all zones.